src/features/session_replay/aggregate/index.js

```diff
@@ -0,0 +1,292 @@
+    this.isFirstChunk = false
+    /** Payload metadata -- Should indicate that the payload being sent has a full DOM snapshot. */
+    this.hasSnapshot = false
+    /** Payload metadata -- Should indicate that the payload being sent contains an error. */
+    this.hasError = false
+
+    /** A value which increments with every new mutation node reported. Resets after a harvest is sent */
+    this.payloadBytesEstimation = 0
+
+    /** The method to stop recording. This defaults to a noop, but is overwritten once the recording library is imported and initialized */
+    this.stopRecording = () => { /* no-op until set by rrweb initializer */ }
+
+    // The SessionEntity class can emit a message indicating the session was cleared and reset (expiry, inactivity). This feature must abort and never resume if that occurs.
+    this.ee.on('session-reset', () => {
+      this.abort()
+    })
+
+    // The SessionEntity class can emit a message indicating the session was paused (visibility change). This feature must stop recording if that occurs.
+    this.ee.on(SESSION_EVENTS.PAUSE, () => { this.stopRecording() })
+    // The SessionEntity class can emit a message indicating the session was resumed (visibility change). This feature must start running again (if already running) if that occurs.
+    this.ee.on(SESSION_EVENTS.RESUME, () => {
+      if (!this.initialized || this.mode === MODE.OFF) return
+      this.startRecording()
+      this.takeFullSnapshot()
+    })
+
+    // Bespoke logic for new endpoint.  This will change as downstream dependencies become solidified.
+    this.scheduler = new HarvestScheduler('blob', {
+      onFinished: this.onHarvestFinished.bind(this),
+      retryDelay: this.harvestTimeSeconds,
+      getPayload: this.prepareHarvest.bind(this),
+      // TODO -- this stuff needs a better way to be handled
+      includeBaseParams: false,
+      customUrl: 'https://vortex-alb.stg-single-tooth.cell.us.nr-data.net/blob',
+      raw: true,
+      gzip: true
+    }, this)
+
+    // Wait for an error to be reported.  This currently is wrapped around the "Error" feature.  This is a feature-feature dependency.
+    // This was to ensure that all errors, including those on the page before load and those handled with "noticeError" are accounted for. Needs evalulation
+    registerHandler('err', (e) => {
+      this.hasError = true
+      // run once
+      if (this.errorNoticed) return
+      // if the error was noticed AFTER the recorder was imported....
+      if (this.mode === MODE.ERROR && !!recorder) {
+        this.errorNoticed = true
+        this.scheduler.runHarvest({ needResponse: true })
+        this.stopRecording()
+        this.mode = MODE.FULL
+        this.startRecording()
+        this.scheduler.startTimer(this.harvestTimeSeconds)
+
+        const { session } = getRuntime(this.agentIdentifier)
+        session.write({ ...session.state, sessionReplay: this.mode })
```

Comment:
`session.write` happens 3 times in this file; is it necessary if the session itself (should) handles writing `agentRuntime` to storage on user nav away?

The point of the intermediary local mem is to be recipient of all the modification and then flushing it once to long term storage at EoL.

Reply:
good call, makes sense.  will just update state of local module, which will sync its state to local storage when the time comes (ie vis change)